compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.cpp

```diff
@@ -464,6 +464,69 @@ static void printWorkgroupCountRegion(OpAsmPrinter &p, Operation *op,
                 /*printBlockTerminators=*/true);
 }
 
+//===----------------------------------------------------------------------===//
+// custom<OptionalWorkgroupCountRegion>($body)
+//===----------------------------------------------------------------------===//
+
+static ParseResult parseOptionalWorkgroupCountRegion(OpAsmParser &parser,
+                                                     Region &body) {
+  // Empty region if not specified.
+  if (failed(parser.parseOptionalKeyword("count"))) {
+    return success();
+  }
+
+  SmallVector<OpAsmParser::Argument> args;
+  if (failed(parser.parseArgumentList(args, AsmParser::Delimiter::Paren,
+                                      /*allowType=*/true,
+                                      /*allowAttrs=*/true))) {
+    return failure();
+  }
+
+  // Return types must be 3 dimensions (workgroup count XYZ).
+  SmallVector<Type> returnTypes;
+  if (failed(parser.parseArrowTypeList(returnTypes))) {
+    return failure();
+  }
+  if (returnTypes.size() != 3 ||
+      !llvm::all_of(returnTypes, [](Type type) { return type.isIndex(); })) {
+    return parser.emitError(parser.getCurrentLocation())
+           << "workgroup count region must return the XYZ dimension counts";
+  }
+
+  // Parse region contents.
+  if (failed(parser.parseRegion(body, args, /*enableNameShadowing=*/false))) {
+    return failure();
+  }
+
+  // Verify the return types match.
+  for (auto returnOp : body.getOps<IREE::HAL::ReturnOp>()) {
+    for (auto [resultType, returnType] :
+         llvm::zip_equal(returnTypes, returnOp.getOperandTypes())) {
+      if (resultType != returnType) {
+        return returnOp.emitOpError()
+               << "operands do not match expected region return types";
```

Comment:
I would expect this to be a verifier error rather than a parser failure.

Reply:
cleaned it up a bit and added some additional verification!